pkg/etcdcli/health.go

```diff
@@ -0,0 +1,154 @@
+			ctx, cancel := context.WithCancel(context.Background())
+			// linearized request to verify health of member
+			_, err = cli.Get(ctx, "health")
+			cancel()
+			hc := healthCheck{Member: member, Healthy: false, Started: true, Took: time.Since(st).String()}
+			if err == nil || err == rpctypes.ErrPermissionDenied {
+				hc.Healthy = true
+			} else {
+				hc.Error = fmt.Errorf("health check failed: %w", err)
+			}
+			hch <- hc
+		}(member)
+	}
+
+	wg.Wait()
+	close(hch)
+
+	for healthCheck := range hch {
+		memberHealth = append(memberHealth, healthCheck)
+	}
+
+	return memberHealth
+}
+
+// Status returns a reporting of memberHealth status
+func (h memberHealth) Status() string {
+	healthyMembers := h.GetHealthyMemberNames()
+
+	status := []string{}
+	if len(h) == len(healthyMembers) {
+		status = append(status, fmt.Sprintf("%d members are available", len(h)))
+	} else {
+		status = append(status, fmt.Sprintf("%d of %d members are available", len(healthyMembers), len(h)))
+		for _, etcd := range h {
+			switch {
+			case !etcd.Started:
+				status = append(status, fmt.Sprintf("%s has not started", GetMemberNameOrHost(etcd.Member)))
+				continue
+			case !etcd.Healthy:
+				status = append(status, fmt.Sprintf("%s is unhealthy", etcd.Member.Name))
+				continue
+			}
+		}
+	}
+	return strings.Join(status, ", ")
+}
+
+// Healthy returns healthy members
+func (h memberHealth) GetHealthyMembers() []*etcdserverpb.Member {
+	members := []*etcdserverpb.Member{}
+	for _, etcd := range h {
+		if etcd.Healthy {
+			members = append(members, etcd.Member)
+		}
+	}
+	return members
+}
+
+// HealthyMemberNames returns a list of healthy member names
+func (h memberHealth) GetHealthyMemberNames() []string {
+	memberNames := []string{}
+	for _, etcdMember := range h.GetHealthyMembers() {
+		memberNames = append(memberNames, etcdMember.Name)
+	}
+	return memberNames
+}
+
+// Unhealthy returns unhealthy members
+func (h memberHealth) GetUnhealthyMembers() []*etcdserverpb.Member {
+	members := []*etcdserverpb.Member{}
+	for _, etcd := range h {
+		if !etcd.Healthy {
+			members = append(members, etcd.Member)
+		}
+	}
+	return members
+}
+
+// UnhealthyMemberNames returns a list of unhealthy member names
+func (h memberHealth) GetUnhealthyMemberNames() []string {
+	memberNames := []string{}
+	for _, etcdMember := range h.GetUnhealthyMembers() {
+		if len(etcdMember.ClientURLs) == 0 {
+			memberNames = append(memberNames, GetMemberNameOrHost(etcdMember))
+			continue
+		}
+		memberNames = append(memberNames, etcdMember.Name)
+	}
+	return memberNames
+}
+
+// Unstarted returns unstarted members
+func (h memberHealth) GetUnstartedMembers() []*etcdserverpb.Member {
+	members := []*etcdserverpb.Member{}
+	for _, etcd := range h {
+		if !etcd.Started {
+			members = append(members, etcd.Member)
+		}
+	}
+	return members
+}
+
+// UnstartedMemberNames returns a list of unstarted member names
+func (h memberHealth) GetUnstartedMemberNames() []string {
```

Comment:
I think all the *Name variant functions could be replaced by a function that generically accepts `[]*etcdserverpb.Member` and returns `[]string` of their names

Reply:
well we can't conclude health from only the member and I would like to only perform a  single series of health checks so names function will have the following signature.
```
 func GetHealthyMemberNames(memberHealth []healthCheck) []string {
```

Reply:
but then I wonder why no keep them as a method?